roomserver/storage/shared/room_updater.go

```diff
@@ -0,0 +1,262 @@
+package shared
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/matrix-org/dendrite/roomserver/types"
+	"github.com/matrix-org/gomatrixserverlib"
+)
+
+type RoomUpdater struct {
+	transaction
+	d                       *Database
+	roomInfo                *types.RoomInfo
+	latestEvents            []types.StateAtEventAndReference
+	lastEventIDSent         string
+	currentStateSnapshotNID types.StateSnapshotNID
+}
+
+func rollback(txn *sql.Tx) {
+	if txn == nil {
+		return
+	}
+	txn.Rollback() // nolint: errcheck
+}
+
+func NewRoomUpdater(ctx context.Context, d *Database, txn *sql.Tx, roomInfo *types.RoomInfo) (*RoomUpdater, error) {
```

Comment:
This leaks `sql.Tx` outside the storage layer, no?

Reply:
Well, sort of, but this is `shared.NewRoomUpdater` rather than `(storage.Database).GetRoomUpdater()` — it's this way because `NewRoomUpdater` needs to be exported to the `sqlite3` package so that it can override the `GetRoomUpdater` and `MembershipUpdater` functions with ones that pass in a `nil` txn.

FWIW the membership and latest events updater have always been this way.